File: backends/ruby-hanami/app/usecase/auth/interactor.rb
# frozen_string_literal: true
#
# 認証ユースケースのインターフェースを定義するモジュール。
#
# @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

module UseCase
  module Auth
    # 認証に関するユースケースのインターフェースです。
    # @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
    class Interactor
      # @param staff_repo [Domain::Staff::Repository] スタッフリポジトリ
      # @param invitation_auth_repo [Domain::Invitation::AuthRepository] 招待認証キャッシュリポジトリ
      def initialize(staff_repo, invitation_auth_repo)
        @staff_repo           = staff_repo
        @invitation_auth_repo = invitation_auth_repo
      end

      def find_user(id)
        entity = @staff_repo.find_by_id(id)
        raise "staff_not_found" unless entity
        Domain::Staff::Vo.new(id: entity.id, name: entity.name, avatar: entity.avatar, role: entity.role)
      end

      def login(dto)
        entity = @staff_repo.find_by_provider(dto.provider, dto.provider_id)
        now = Time.now

        if entity
          updated               = entity.dup
          updated.avatar        = dto.avatar
          updated.last_login_at = now
          updated.updated_at    = now
          saved = @staff_repo.save(updated)
        else
          token = dto.invitation_token
          if token.nil? || token.empty? || @invitation_auth_repo.find(token).nil?
            raise Domain::ForbiddenError.new("invitation_required")
          end
          @invitation_auth_repo.remove(token)
          new_entity = Domain::Staff::Entity.new(
            id:            nil,
            name:          dto.name,
            email:         dto.email,
            provider:      dto.provider,
            provider_id:   dto.provider_id,
            avatar:        dto.avatar,
            role:          Domain::Staff::Role::MEMBER,
            last_login_at: now,
            created_at:    now,
            created_by:    0,
            updated_at:    now,
            updated_by:    0,
            deleted_at:    nil,
            deleted_by:    nil,
            version:       1,
          )
          saved = @staff_repo.save(new_entity)
        end

        Domain::Staff::Vo.new(id: saved.id, name: saved.name, avatar: saved.avatar, role: saved.role)
      end
    end
  end
end

File: backends/go-beego/internal/usecase/auth/interactor.go
// Package auth は認証ユースケースを提供します。
package auth

import (
	dominvitation "authorization-go-beego/internal/domain/invitation"
	domstaff "authorization-go-beego/internal/domain/staff"
	"authorization-go-beego/pkg/apperror"
	"time"
)

// Interactor は認証ユースケースの実装です。
type Interactor struct {
	staffRepo          domstaff.Repository
	invitationAuthRepo dominvitation.AuthRepository
}

// NewInteractor は Interactor を生成します。
func NewInteractor(staffRepo domstaff.Repository, invitationAuthRepo dominvitation.AuthRepository) *Interactor {
	return &Interactor{staffRepo: staffRepo, invitationAuthRepo: invitationAuthRepo}
}

// FindUser はスタッフIDでスタッフ情報を取得します。
func (uc *Interactor) FindUser(dto FindUserDto) (*domstaff.Vo, error) {
	s, err := uc.staffRepo.FindByID(&domstaff.Staff{ID: dto.ID})
	if err != nil {
		return nil, err
	}
	if s == nil {
		return nil, apperror.NotFound("user_not_found")
	}
	return staffToVo(s), nil
}

// Login はGoogle OAuthのプロバイダ情報でスタッフを登録またはログインします。
func (uc *Interactor) Login(dto LoginDto) (*domstaff.Vo, error) {
	existing, err := uc.staffRepo.FindByProvider(&domstaff.Staff{Provider: dto.Provider, ProviderID: dto.ProviderID})
	if err != nil {
		return nil, err
	}

	now := time.Now()
	if existing == nil {
		if dto.InvitationToken == "" {
			return nil, apperror.Forbidden("invitation_required")
		}
		found, err := uc.invitationAuthRepo.Find(dto.InvitationToken)
		if err != nil {
			return nil, err
		}
		if found == "" {
			return nil, apperror.Forbidden("invitation_required")
		}
		if err := uc.invitationAuthRepo.Remove(dto.InvitationToken); err != nil {
			return nil, err
		}

		zero := uint(0)
		newStaff := &domstaff.Staff{
			Name:        dto.Name,
			Email:       dto.Email,
			Provider:    dto.Provider,
			ProviderID:  dto.ProviderID,
			Avatar:      dto.Avatar,
			Role:        domstaff.RoleMember,
			LastLoginAt: &now,
			CreatedAt:   now,
			CreatedBy:   &zero,
			UpdatedAt:   now,
			UpdatedBy:   &zero,
		}
		saved, err := uc.staffRepo.Save(newStaff)
		if err != nil {
			return nil, err
		}
		return staffToVo(saved), nil
	}

	existing.Avatar = dto.Avatar
	existing.LastLoginAt = &now
	existing.UpdatedAt = now
	saved, err := uc.staffRepo.Save(existing)
	if err != nil {
		return nil, err
	}
	return staffToVo(saved), nil
}

// staffToVo はスタッフエンティティをVOに変換します。
func staffToVo(s *domstaff.Staff) *domstaff.Vo {
	return &domstaff.Vo{
		ID:     s.ID,
		Name:   s.Name,
		Avatar: s.Avatar,
		Role:   s.Role,
	}
}
